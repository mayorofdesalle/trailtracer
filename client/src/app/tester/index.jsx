import '@testing-library/jest-dom';

import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { BrowserRouter } from 'react-router-dom';

import './mocks/i18n.mock';
import './mocks/matchMedia.mock';
import './mocks/resizeObserver.mock';

import store from '@app/store';
import GlobalStyle from '@components/style/GlobalStyle';
import Theme from '@components/style/Theme';

const Providers = ({ children }) => {
    return (
        <Provider store={store}>
            <Theme>
                <GlobalStyle />
                <BrowserRouter>
                    {children}
                </BrowserRouter>
            </Theme>
        </Provider>
    );
};

const customRender = (ui, options) =>
    render(ui, { wrapper: Providers, ...options });

// re-export everything
export * from '@testing-library/react';
export * from 'vitest';

// override render method
export { customRender as render };
