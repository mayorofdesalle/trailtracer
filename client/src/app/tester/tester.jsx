import { render } from '@testing-library/react'
import { Provider } from 'react-redux';

import './mocks/matchMedia.mock';
import store from '@app/store';
import Theme from '@components/style/Theme'

const Providers = ({ children }) => {
    return (
        <Provider store={store}>
            <Theme>
                {children}
            </Theme>
        </Provider>
    );
};

const customRender = (ui, options) =>
    render(ui, { wrapper: Providers, ...options });

// re-export everything
export * from '@testing-library/react';
export * from 'vitest';

// override render method
export { customRender as render };