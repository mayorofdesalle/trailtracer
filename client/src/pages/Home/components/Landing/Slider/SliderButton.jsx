import { useRef } from 'react';
import Draggable from 'react-draggable';
import styled from 'styled-components';
import PropTypes from 'prop-types';

import Button from '@components/ui/Button';
import Icon from '@components/ui/Icon';

/**
 * SliderButtonInner
 * @description
 * This component is a styled Button component creating the base for the SliderButton.
 **/
const SliderButtonInner = styled(Button)`
    position: absolute;
    display: flex;
    height: 60%;
    width: 15%;
    padding: 0;
    color: ${({ theme }) => theme.colors.text};
    background-color: ${({ theme }) => theme.colors.secondary};
    cursor: move;
    touch-action: pan-y;
    z-index: 1;
    transition: 200ms ease-in-out;

    &.isDragging {
        background-color: ${({ theme }) => theme.colors.primary};
        transition: none;
    }

    &.isDragging > * {
        fill: ${({ theme }) => theme.colors.tertiary};
    }

    & > svg {
        pointer-events: none;
    }

    & > #arrow-left-fill {
        transform: translateX(0.3rem);
    }

    & > #arrow-right-fill {
        transform: translateX(-0.3rem);
    }

    &:hover > #arrow-left-fill, &.isDragging > #arrow-left-fill {
        transform: translateX(0.1rem);
    }

    &:hover > #arrow-right-fill, &.isDragging > #arrow-right-fill {
        transform: translateX(-0.1rem);
    }
`;

/**
 * SliderButton
 * @param {Number} width - The width of the slider
 * @description
 * This component is a draggable button that is used to slide between pages.
 * It uses the Draggable component from react-draggable to create the button.
 * The button is snapped to the closest page when released.
 **/
const SliderButton = ({ onDrag, onStop, position }) => {
    const button = useRef();

    return (
        <Draggable
            defaultClassNameDragging='isDragging'
            position={{ x: position, y: 0 }}
            nodeRef={button}
            axis='x'
            bounds='parent'
            onDrag={(e, button) => onDrag(e, button)}
            onStop={(e, button) => onStop(e, button)}
        >
            <SliderButtonInner ref={button}>
                <Icon name='arrow-left-fill' />
                <Icon name='arrow-right-fill' />
            </SliderButtonInner>
        </Draggable>
    );
};

SliderButton.propTypes = {
    onDrag: PropTypes.func.isRequired,
    onStop: PropTypes.func.isRequired,
    position: PropTypes.number
};

export default SliderButton;