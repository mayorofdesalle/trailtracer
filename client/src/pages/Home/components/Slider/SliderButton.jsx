import { useRef, useState } from 'react';
import { useDispatch } from 'react-redux';
import Draggable from 'react-draggable';
import styled from 'styled-components';
import PropTypes from 'prop-types';

import { modify } from '@features/background/backgroundSlice';
import Button from '@components/ui/Button';
import Icon from '@components/ui/Icon';
import { fadeIn } from '@components/misc/anims';

/**
 * function findClosestPage(pagePositions, position)
 * @param {Array} pagePositions - The positions of the pages
 * @param {Number} position - The current position of the button
 * @returns {Number} The index of the closest page
 * @description
 * This function finds the closest page to the current position of the button.
 **/
const findClosestPage = (pagePositions, position) => {
    const closest = pagePositions.reduce((prev, curr) => {
        return Math.abs(curr - position) < Math.abs(prev - position) ? curr : prev;
    });

    return pagePositions.indexOf(closest);
};

/**
 * function mapPositionToPage(position, sliderWidth, wrapperWidth)
 * @param {Number} position - The current position of the button
 * @param {Number} sliderWidth - The width of the Slider
 * @param {Number} wrapperWidth - The width of the ScrollWrapper
 * @returns {Number} The position of the page
 * @description
 * This function maps the position of the button to the position of the page.
 * (number - inMin) * (outMax - outMin) / (inMax - inMin) + outMin;
 * 3 * wrapperWidth for 4 bentos since the scroll is set from the left.
 **/
const mapPositionToPage = (position, sliderWidth, wrapperWidth) => {
    return (position - 3 * sliderWidth / 8) * (3 * wrapperWidth) / (6 * sliderWidth / 8) + (3 * wrapperWidth);
};

const adjustBackground = (position, pagePosition, dispatch) => {
    const transitionTime = 300; // milliseconds
    const distance = pagePosition - position;

    const easeInOutSine = (t) => {
        return -(Math.cos(Math.PI * t) - 1) / 2;
    };

    const adjust = (repeats = transitionTime) => {
        if (repeats > 0) {
            const animTime = easeInOutSine(repeats / transitionTime);

            const delta = distance * animTime / transitionTime;

            dispatch(modify({ layers: 0, distortion: 0.001 * delta, slope: 0.001 * delta, jitter: 0.001 * delta }));
            setTimeout(() => adjust(repeats - 1), 1);
        }
    };

    adjust();
};

/**
 * ButtonBase
 * @description
 * This component is a styled Button component creating the base for the SliderButton.
 **/
const ButtonBase = styled(Button)`
    position: absolute;
    display: flex;
    height: 60%;
    width: 15%;
    padding: 0;
    fill: ${({ theme }) => theme.colors.text};
    background-color: ${({ theme }) => theme.colors.secondary};
    cursor: move;
    touch-action: pan-y;
    z-index: 1;
    transition: 300ms ease-in-out;
    animation: ${fadeIn} 1s ease-in-out;

    & > #arrow-left-fill {
        transform: translateX(0.3rem);
        transition: 200ms ease-in-out;
    }

    & > #arrow-right-fill {
        transform: translateX(-0.3rem);
        transition: 200ms ease-in-out;
    }

    &:hover > #arrow-left-fill {
        transform: translateX(0.1rem);
    }

    &:hover > #arrow-right-fill {
        transform: translateX(-0.1rem);
    }
`;

/**
 * SliderButton
 * @param {Number} width - The width of the slider
 * @description
 * This component is a draggable button that is used to slide between pages.
 * It uses the Draggable component from react-draggable to create the button.
 * The button is snapped to the closest page when released.
 **/
const SliderButton = ({ wrapperWidth, sliderWidth, setPageScroll }) => {
    const pagePositions = [-3 * sliderWidth / 8, -sliderWidth / 8, sliderWidth / 8, 3 * sliderWidth / 8];
    const [currentPage, setCurrentPage] = useState(0);
    const button = useRef();
    const dispatch = useDispatch();

    return (
        <Draggable
            defaultClassNameDragging='isDragging'
            position={{ x: pagePositions[currentPage], y: 0 }}
            nodeRef={button}
            axis='x'
            bounds='parent'
            onStop={(e, button) => {
                const closestPage = findClosestPage(pagePositions, button.x);
                setCurrentPage(closestPage);
                setPageScroll(mapPositionToPage(pagePositions[closestPage], sliderWidth, wrapperWidth));
                adjustBackground(button.x, pagePositions[closestPage], dispatch);
            }}
            onDrag={(e, button) => {
                setPageScroll(mapPositionToPage(button.x, sliderWidth, wrapperWidth));
                dispatch(modify({ layers: 0, distortion: 0.001 * button.deltaX, slope: 0.001 * button.deltaX, jitter: 0.001 * button.deltaX }));
            }}
        >
            <ButtonBase ref={button}>
                <Icon name='arrow-left-fill' height={4} width={4} />
                <Icon name='arrow-right-fill' height={4} width={4} />
            </ButtonBase>
        </Draggable>
    );
};

SliderButton.propTypes = {
    wrapperWidth: PropTypes.number,
    sliderWidth: PropTypes.number.isRequired,
    setPageScroll: PropTypes.func.isRequired,
};

export default SliderButton;