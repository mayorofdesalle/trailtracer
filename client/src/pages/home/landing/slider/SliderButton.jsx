import PropTypes from 'prop-types';
import { useRef } from 'react';
import Draggable from 'react-draggable';
import styled from 'styled-components';

import { Button } from '@components/ui';
import { Icon } from '@components/ui';

const SliderButtonInner = styled(Button)`
    position: absolute;
    display: flex;
    height: 60%;
    width: 15%;
    padding: 0;
    color: ${({ theme }) => theme.colors.text};
    background-color: ${({ theme }) => theme.colors.secondary};
    cursor: move;
    touch-action: pan-y;
    z-index: 1;
    transition: 200ms ease-in-out;

    & > svg {
        transition: inherit;
        width: 100%;
        height: 100%;
    }

    & > #arrow-left-fill {
        transform: translateX(0.3rem);
    }

    & > #arrow-right-fill {
        transform: translateX(-0.3rem);
    }

    &.isDragging {
        background-color: ${({ theme }) => theme.colors.primary};

        & > * {
            transition: 200ms ease-in-out;
            fill: ${({ theme }) => theme.colors.tertiary};
        }

        transition: none;
    }

    &.isDragging, &:hover {
        
        & > #arrow-left-fill {
            transform: translateX(0.1rem);
        }

        & > #arrow-right-fill {
            transform: translateX(-0.1rem);
        }
    }
`;

/**
 * SliderButton
 * @param {Function} onDrag - The function to call when the button is dragged.
 * @param {Function} onStop - The function to call when the button is released.
 * @param {Number} position - The position of the button.
 * @description
 * Styled react-draggable button that controls the scroll amount.
 **/
const SliderButton = ({ onDrag, onStop, position, ...props }) => {
    const button = useRef();

    return (
        <Draggable
            defaultClassNameDragging='isDragging'
            position={{ x: position, y: 0 }}
            nodeRef={button}
            axis='x'
            bounds='parent'
            onDrag={(e, button) => onDrag(e, button)}
            onStop={(e, button) => onStop(e, button)}
        >
            <SliderButtonInner ref={button} {...props}>
                <Icon name='arrow-left-fill' />
                <Icon name='arrow-right-fill' />
            </SliderButtonInner>
        </Draggable>
    );
};

SliderButton.propTypes = {
    onDrag: PropTypes.func.isRequired,
    onStop: PropTypes.func.isRequired,
    position: PropTypes.number
};

export default SliderButton;