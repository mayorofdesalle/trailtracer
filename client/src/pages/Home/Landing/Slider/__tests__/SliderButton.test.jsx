import { describe, expect, it, render, screen, vi } from '@app/tester';

import SliderButton from '../SliderButton';

describe('SliderButton', () => {

    it('should render correctly', () => {
        render(<SliderButton onDrag={vi.fn()} onStop={vi.fn()} />);
        const button = screen.getByRole('button');

        expect(button).toBeInTheDocument();
        expect(button.firstChild).toBeInTheDocument();
        expect(button.firstChild).toHaveRole('img');
        expect(button.lastChild).toBeInTheDocument();
        expect(button.lastChild).toHaveRole('img');
    });

    it('should match snapshot', () => {
        const { asFragment } = render(<SliderButton onDrag={vi.fn()} onStop={vi.fn()} />);
        expect(asFragment()).toMatchSnapshot();
    });

});