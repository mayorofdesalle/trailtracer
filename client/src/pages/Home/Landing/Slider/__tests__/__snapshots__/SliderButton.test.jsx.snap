// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SliderButton > should match snapshot 1`] = `
<DocumentFragment>
  <button
    class="sc-beySPh sc-irLvIq zbkUW fEWcSo react-draggable"
    style="transform: translate(undefinedpx,0px);"
    type="button"
  >
    <svg
      class="sc-jTQCzO ldKJNf"
      fill="currentColor"
      role="img"
    >
      <title>
        arrow-left-fill
      </title>
      <use
        href="http://localhost:3000/src/assets/icons.svg#arrow-left-fill"
      />
    </svg>
    <svg
      class="sc-jTQCzO ldKJNf"
      fill="currentColor"
      role="img"
    >
      <title>
        arrow-right-fill
      </title>
      <use
        href="http://localhost:3000/src/assets/icons.svg#arrow-right-fill"
      />
    </svg>
  </button>
</DocumentFragment>
`;
